### PR TITLE
Adapt single currency settings page UI to match WCPay theme

### LINKED ISSUE #2874: Update single currency settings UI to match WCPay settings
Follow-up issue of #2745 

Update the MC single currency settings page to use the same layout/style/functionality as the new 2 column payments grouped settings page. Specs in paJDYF-1gf-p2

Screenshot:

image

### PR DESCRIPTION
**Pull request by tpaksu**
Fixes #2874

#### Changes proposed in this Pull Request

Create a React rendered component for single currency settings

This PR provides an UI for the single currency settings compatible with the new WCPay grouped settings interface, which enables users see the current rate of a currency, when it's last updated, set manual rates for them, set price rounding fraction, set price charm options, and test their settings instantly with a preview section, which has the field for setting a base amount and see the converted amount regarding the new settings instantly.

The new screen looks like this:

image

One note; Since the main currency settings will be updated on #2832, the header part couldn't be removed yet, it'll be done when merging this with #2832 to the `develop` branch.

And one nitpick; I needed to use a context provider for both currency selection screen and single currency screen to exchange display statuses and open/close screen methods like this:

```jsx
<MultiCurrencySettingsContext.Provider
    value={ {
        isSingleCurrencyScreenOpen: isSingleCurrencyScreenOpen,
        currencyCodeToShowSettingsFor: currencyCodeToShowSettingsFor,
        openSingleCurrencySettings: handleOpenSingleCurrencySettings,
        closeSingleCurrencySettings: handleCloseSingleCurrencySettings,
    } }
>
    { ! isSingleCurrencyScreenOpen ? (
        <EnabledCurrencies />
    ) : (
        <SingleCurrencySettings />
    ) }
</MultiCurrencySettingsContext.Provider>
```

Taking this into account, it can be easily merged with #2832 and be ready to use.

##### Automatic testing

- Run `npm run test:js single-currency-settings``

##### Manual testing

- Open *WooCommerce > Settings* and click to *Multi-Currency* tab
- Add some currencies to the list if the list is empty
- Click *Manage* on any currency
- See if the default settings show as 'Automatic' as the exchange rate selection, 'None' for the price charm, and '1.00' as the rounding mode for non-zero-decimal currencies, and '100' for the zero-decimal ones.
- Change some settings and save the changes
- Go back to the main screen with the **Enabled Currencies** link on the breadcrumb (I offered to add a back button to the screen), and @LevinMedia told me that it's an open issue for the Wizard component which is waiting on hold for now, and there's no need to add a separate one for this yet.)
- Open another currency settings, and check if the settings are correct, not displaying the previous one's settings
- Refresh the page and check if the previously set settings are correctly displayed
- Change the settings and see if the Preview div is updated correctly for the given settings

-------------------

- [x] Added changelog entry (or does not apply)
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

**Post merge**

- [ ] Added testing instructions to the [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) (or does not apply)

### CODE REVIEW

**Comment by tpaksu** [comment]
The construction of the single-currency page is done, supported by a dummy integration with the MC settings page, but I need some help about getting fresh data from the API everytime the single currency settings component is rendered, currently it seems to memoize the previous currencies and won't refresh the data from the server when the user opens the same currency again. This causes the single currency page showing the previous currency's settings if the currency has been clicked more than once. 

For example this scenario:
1) Open MC settings page,
2) Click 'manage' to CAD currency and do some changes.
3) Save the changes
4) Click on the 'JPY' currency
5) Ensure it has different settings than what you did for CAD
6) Click 'manage' again for CAD, it'll show the settings of JPY.

Do you have any ideas to prevent this? /cc @ismaeldcom @jessepearson

**Comment by ismaeldcom** [comment]
@tpaksu You're dealing here with the `@wordpress/data` resolvers way of work and cache. You are saving the currency settings on the store like this:
```js
store: {
...
  currency_settings: {
    price_rounding:"0.25"
    price_charm:"0"
    ...
```
That data got replaced when the selector is called every time you access a single currency settings page. But the selector have a cache, so if you go to EUR, then to USD and again to EUR, you'll end with the USD settings. Because the resolver will not be called again for the same request.

I can think of two options here, you can invalidate the cache to ensure the resolvers got called again. Or IMO a better approach would be storing the settings for each currency, so instead of the structure above you can use something like:
```js
store: {
...
  currency_settings: {
    EUR: {
      price_rounding:"0.25"
      price_charm:"0"
      ....
    },
    USD: {
    ....
```
In this way, the selector will not overwrite the settings, and we'll be only doing one request per currency rather than one every time the currency settings page is acceded. Taking advantage from the resolvers cache.

BTW If you don't have it, it's good to have the [Redux DevTools browser extension](https://chrome.google.com/webstore/detail/redux-devtools/lmhkpmbekcpmknklioeibfkpmmfibljd?hl=en), so you can watch every action dispatched and state change.

**Comment by tpaksu** [comment]
> Or IMO a better approach would be storing the settings for each currency, so instead of the structure above you can use something like:

> store: {
...
  currency_settings: {
    EUR: {
      price_rounding:"0.25"
      price_charm:"0"
      ....
    },
    USD: {
    ....
> In this way, the selector will not overwrite the settings, and we'll be only doing one request per currency rather than one every time the currency settings page is acceded. Taking advantage from the resolvers cache.

Yes, this worked! Thanks for recommending it, it's way cleaner and manageable than whatever I did before to solve it :) 

> BTW If you don't have it, it's good to have the Redux DevTools browser extension, so you can watch every action dispatched and state change.

I'll install it now, definitely will be better than good ol' `console.log` :)

File: includes/multi-currency/RestController.php
Comment (dmallory42) [inline includes/multi-currency/RestController.php:120]: I'm wondering if perhaps here it may be better to check the list of `available_currencies` from `MultiCurrency` to make sure the currency is valid, rather than validating that it is 3 characters? If it's not feasible to do that easily from within the context of the `RestController`, I think this approach wouldn't an issue, however.

**Review by dmallory42 — CHANGES_REQUESTED** [review]
I made it 😅

This is looking really good, @tpaksu! A few things I noticed while testing to consider which I'll mention below, but other than that I think this is pretty much good to go.

I noticed that clicking on the "Learn more" links in the Formatting Rules section currently has no effect (not sure if this is in the scope of the PR but thought I would mention):

Screenshot 2021-09-14 at 14 41 10

I noticed that when I change the currency rate (e.g. switch from automatic to manual and change the rate), save the change, and then go back to the main Enabled Currencies page via the breadcrumb link, the rate displayed in the main table containing all of the enabled currencies does not update for the currency. I have to refresh the page to get this rate to update. Is it possible to make this update once the change is saved on the single currency settings page?

Also, I wonder whether it would look better if the "Current Rate..." text here lined up with the "Fetch rates automatically" text above it. This may be more one for @LevinMedia to give feedback on, or something that has been discussed already so feel free to ignore in that case! 

Screenshot 2021-09-14 at 12 27 41

Let me know if you need any more info on anything I mentioned! 🙂

File: includes/multi-currency/RestController.php
Comment (tpaksu) [reply includes/multi-currency/RestController.php:120]: Thanks @dmallory42 for the detailed review, I knew you would make it 🚀 😁 Here are my answers:

>I noticed that clicking on the "Learn more" links in the Formatting Rules section currently has no effect (not sure if this is in the scope of the PR but thought I would mention):

I couldn't find any links for them, since the old version doesn't contain them and the new version I copied from Figma's that @LevinMedia provided, so I don't know where they should navigate the user.

> I noticed that when I change the currency rate (e.g. switch from automatic to manual and change the rate), save the change, and then go back to the main Enabled Currencies page via the breadcrumb link, the rate displayed in the main table containing all of the enabled currencies does not update for the currency. I have to refresh the page to get this rate to update. Is it possible to make this update once the change is saved on the single currency settings page?

Yep, I know that, this is a result of me and @daquinons doing the same screen split by two, I didn't want to touch that part because it's been re-designed by @daquinons, and that'll be a concern to solve when the two PR's get merged. Does it make sense?

> I wonder whether it would look better if the "Current Rate..." text here lined up with the "Fetch rates automatically" text above it. This may be more one for @LevinMedia to give feedback on, or something that has been discussed already so feel free to ignore in that case!

Yep, I'm also uncomfortable by that look, but that's how the Figma's are. 7KVvtMQgJxVAUcPPxTB29r-fi-174%3A3922, we should ask @LevinMedia about that part too.

Hope all does make sense.

**Comment by tpaksu** [comment]
Thanks @dmallory42 for the detailed review, I knew you would make it 🚀 😁 Here are my answers:

>I noticed that clicking on the "Learn more" links in the Formatting Rules section currently has no effect (not sure if this is in the scope of the PR but thought I would mention):

I couldn't find any links for them, since the old version doesn't contain them and the new version I copied from Figma's that @LevinMedia provided, so I don't know where they should navigate the user.

> I noticed that when I change the currency rate (e.g. switch from automatic to manual and change the rate), save the change, and then go back to the main Enabled Currencies page via the breadcrumb link, the rate displayed in the main table containing all of the enabled currencies does not update for the currency. I have to refresh the page to get this rate to update. Is it possible to make this update once the change is saved on the single currency settings page?

Yep, I know that, this is a result of me and @daquinons doing the same screen split by two, I didn't want to touch that part because it's been re-designed by @daquinons, and that'll be a concern to solve when the two PR's get merged. Does it make sense?

> I wonder whether it would look better if the "Current Rate..." text here lined up with the "Fetch rates automatically" text above it. This may be more one for @LevinMedia to give feedback on, or something that has been discussed already so feel free to ignore in that case!

Yep, I'm also uncomfortable by that look, but that's how the Figma's are. 7KVvtMQgJxVAUcPPxTB29r-fi-174%3A3922, we should ask @LevinMedia about that part too.

Hope all does make sense.

File: includes/multi-currency/RestController.php
Comment (tpaksu) [reply includes/multi-currency/RestController.php]: Fixed in 0523c1e. Thanks again!

**Comment by dmallory42** [comment]
> I couldn't find any links for them, since the old version doesn't contain them and the new version I copied from Figma's that @LevinMedia provided, so I don't know where they should navigate the user.

I think the best places to direct these links to are as follows:

Price Rounding: https://docs.woocommerce.com/document/payments/currencies/multi-currency-setup/#price-rounding
Price Charm: https://docs.woocommerce.com/document/payments/currencies/multi-currency-setup/#price-rounding

> Yep, I know that, this is a result of me and @daquinons doing the same screen split by two, I didn't want to touch that part because it's been re-designed by @daquinons, and that'll be a concern to solve when the two PR's get merged. Does it make sense?

Makes sense, thanks for confirming. Perhaps we should make an issue to track that so we make sure to follow up with a PR after the two PR's are merged.

Thanks for making those updates @tpaksu, I'll approve the PR as it all looks good and as working well, but let's update those "Learn More" links before merging and get the design review. 🙂

**Review by dmallory42 — APPROVED** [review]
All looks good and is working well, great job @tpaksu! 🚀 

There's one small update that needs to be made to add the learn more links (see comment above), but I'm pre-approving this! 

Let's get a design review as well before merging. 🙂

**Comment by tpaksu** [comment]
Updated the links! Thanks @dmallory42!

**Comment by daquinons** [comment]
This looks fantastic, thank you for the hard work @tpaksu. We can coordinate to merge:

- If we merge this one first, I can then rebase my PR and make the needed changes for updating the rates in the MC settings view.
- If we merge both together, we can create an issue to address the rate update issue.

Let me know which one you prefer to proceed! 🚀

**Comment by tpaksu** [comment]
Hi @daquinons, thank you for sharing your thoughts about the merge! I think the first option would be more suitable because it won't require any additional PR's to do the merge, I'm currently waiting for the design review from @LevinMedia to be completed, other than that I don't have anything left for the PR, I suppose. I created a dummy container and context for sharing the data between those components, so it can be easy to run both components in sync. Feel free to apply your implementation to open/close the single currency settings page though :)

**Review by LevinMedia — APPROVED** [review]
My only feedback is on the difference in padding between the top card and the bottom, can you use the padding on the bottom card on the top please? Otherwise LGTM.

**Pull request merged**
